src/interpreter.c

```diff
@@ -166,6 +166,82 @@ static void eval_abstracttype(jl_expr_t *ex, interpreter_state *s)
     if (temp == NULL || !equiv_type(dt, (jl_datatype_t*)jl_unwrap_unionall(temp))) {
         jl_checked_assignment(b, w);
     }
+    dt->incomplete = 0;
+    JL_GC_POP();
+}
+
+static void check_type_completion(jl_datatype_t *bdt, jl_value_t *super, jl_value_t *para)
+{
+    // Check that supertype matches
+    if (!jl_types_equal((jl_value_t*)bdt->super, super))
+        jl_error("Super type does not match forward declaration");
+    // Check that parameters match
+    if (jl_svec_len(para) != jl_svec_len(bdt->parameters)) {
+        jl_error("number of type parameters does not match forward declared type");
+    }
+    for (int i = 0; i < jl_svec_len(para); ++i) {
+        jl_tvar_t *v = (jl_tvar_t*)jl_svecref(para, i);
+        jl_tvar_t *w = (jl_tvar_t*)jl_svecref(bdt->parameters, i);
+        assert(jl_is_typevar(v) && jl_is_typevar(w));
+        if (v->name != w->name)
+            jl_errorf("Name `%s` of type parameter does not match name `%s` from forward declaration.",
+                jl_symbol_name(v->name), jl_symbol_name(w->name));
+        else if (!jl_types_equal(v->lb, w->lb) || !jl_types_equal(v->ub, w->ub)) {
+            jl_errorf("Bounds of type parameter `%s` do not match forward declaration",
+                jl_symbol_name(v->name));
+        }
+    }
+}
+
+static void eval_incompletetype(jl_expr_t *ex, interpreter_state *s)
+{
+    jl_value_t **args = jl_array_ptr_data(ex->args);
+    if (inside_typedef)
+        jl_error("cannot eval a new incomplete type definition while defining another type");
+    jl_value_t *name = args[0];
+    jl_value_t *para = eval_value(args[1], s);
+    jl_value_t *super = NULL;
+    jl_value_t *temp = NULL;
+    jl_datatype_t *dt = NULL;
+    jl_value_t *w = NULL;
+    jl_module_t *modu = s->module;
+    JL_GC_PUSH5(&para, &super, &temp, &w, &dt);
+    assert(jl_is_svec(para));
+    if (jl_is_globalref(name)) {
+        modu = jl_globalref_mod(name);
+        name = (jl_value_t*)jl_globalref_name(name);
+    }
+    assert(jl_is_symbol(name));
+    assert(jl_is_svec(para));
+    jl_binding_t *b = jl_get_binding_wr(modu, (jl_sym_t*)name, 1);
+    if (b->value && jl_is_datatype(jl_unwrap_unionall(b->value))) {
+        super = eval_value(args[2], s);
+        check_type_completion(
+            (jl_datatype_t*)jl_unwrap_unionall(b->value), super, para);
+        return;
+    }
+    dt = jl_new_datatype((jl_sym_t*)name, modu, (jl_datatype_t*)super,
+                         (jl_svec_t*)para, NULL, NULL, 0, 0, 0);
+    w = dt->name->wrapper;
+    temp = b->value;
```

Comment:
this is NULL or the assignment later will fail, per the check a few lines earlier to ensure the result is guaranteed to be compatible with `equiv_type`